connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamingQueryCache.scala

```diff
@@ -0,0 +1,203 @@
+ * no longer active), it is cached for 1 hour so that it is accessible from the client side. It
+ * runs a background thread to run a periodic task that does the following:
+ *   - Check the status of the queries, and drops those that expired (1 hour after being stopped).
+ *   - Keep the associated session active by invoking supplied function `sessionKeepAliveFn`.
+ *
+ * This class helps with supporting following semantics for streaming query sessions:
+ *   - Keep the session and session mapping at connect server alive as long as a streaming query
+ *     is active. Even if the client side has disconnected.
+ *     - This matches how streaming queries behave in Spark. The queries continue to run if
+ *       notebook or job session is lost.
+ *   - Once a query is stopped, the reference and mappings are maintained for 1 hour and will be
+ *     accessible from the client. This allows time for client to fetch status.
+ *     - During this time if the query is restarted (i.e. has a new run id), the reference to
+ *       previous run is dropped. As a result logical query has only the most recent query
+ *       reference cached. This policy can be revisited to cache multiple runs for a query.
+ *   - Note that these semantics are evolving and might change before being finalized in Connect.
+ *   - Future improvements:
+ *     - Provide an API for a users to access session even after they lose session id.
+ *       - Once a user is properly authenticated, the API could return list of sessions that are
+ *         still alive in the connect server for that user.
+ */
+private[connect] class SparkConnectStreamingQueryCache(
+    val sessionKeepAliveFn: (String, String) => Unit, // (userId, sessionId) => Unit.
+    val clock: Clock = new SystemClock(),
+    private val stoppedQueryCachePeriod: Duration = 1.hour, // Configurable for testing.
+    private val sessionPollingPeriod: Duration = 1.minute // Configurable for testing.
+) extends Logging {
+
+  import SparkConnectStreamingQueryCache._
+
+  def registerNewStreamingQuery(sessionHolder: SessionHolder, query: StreamingQuery): Unit = {
+    queryCacheLock.synchronized {
+      val value = QueryCacheValue(
+        userId = sessionHolder.userId,
+        sessionId = sessionHolder.sessionId,
+        session = sessionHolder.session,
+        query = query,
+        expiresAtMs = None)
+
+      queryCache.put(QueryCacheKey(query.id.toString), value) match {
+        case Some(existing) => // Query is being replaced. Can happen when a query is restarted.
+          log.info(
+            s"Replacing existing query in the cache. Query Id: ${query.id}." +
+              s"Existing value $existing, new value $value.")
+        case None =>
+          log.info(s"Adding new query to the cache. Query Id ${query.id}, value $value.")
+      }
+
+      schedulePeriodicChecks() // Starts the scheduler thread if it hasn't started.
```

Comment:
This method also seems to call `queryCacheLock.synchronized {`? Maybe we could remove that line in `schedulePeriodicChecks`?

Reply:
Can do that, but is simpler to just keep the lock. If someone is reading code for this method, it is clear to them that it is locked, without having to look at the call site.

Reply:
Oh I thought doing a lock() inside a lock() would block...? That's not the case in scala...?

Reply:
These are reentrant locks. This is from Java, continues to Scala.